privatesend|wallet|qt: Improve calculations in CreateDenominated/MakeCollateralAmounts

xdustinface opened the pull request:
While the idea of respecting fees in calculations brought up in #3588 and #3640 makes totally sense i still noticed while reviewing/testing the latter that the estimation of just saying use min fee for `148` bytes for inputs and min fee for `34` bytes for outputs and then sum it up isn’t good enough imo.

1. It doen’t respect the general tx overhead
2. It introduces rounding errors with the increasing numbers of inputs/outputs

So even with those estimations we end up in cases where the calculations are slightly off from the expected. This brings inaccuracy to transaction generation and leads to issues (e.g. the. one in `TransactionRecord::decomposeTransaction`).

As solution this PR introduces `CTransactionBuilder` in 13fa395eeb0050225ce2f1a77418d1af96137b0e, a class which wraps transaction creation and allows a simple and failproof creation of transactions with inputs defined by `CompactTallyItem` while also respecting the logics implemented in `CWallet::CreateTransaction`. For output generation it takes care of key generation, provides methods e.g. to add/try to add outputs, get the remaining amount available, and methods to create/commit the transaction to the wallet. To avoid running into the issues described above it does not calculate a seperate fee for the amount of bytes for input/outputs like it was before. Instead it uses bytes for all calculations and calculates a total fee for the transaction when required. This way it’s possible now to calculate the correct amounts/fees used for the created transactions. 

Further on top this PR brings also some small fixes/modifications to the logic behind `CreateDenominated` and `MakeCollateralAmounts` in fba14a11411639d15700e8b658a03b5d0611ff65 and 8e0191ca83b23ff76ca87f8a45daec2a3cd1d74d so please review very carefully there that we can be sure i didn’t mess something here :) 

3d0060165fab3e930a11dae85066d6eb1e36d264 finally makes sure #3597 is not longer an issue. As it now should cover all the cases produced by `MakeCollateralAmounts` correctly.

UdjinM6 commented:
Pls see https://github.com/UdjinM6/dash/commits/pr3657

tl;dr:
1. 209f43f should be reverted cause it violates discard/dustrelay fee definitions
2. I think you messed up "finished" conditions (39a60562aa95ef2a4dbd6a49e00c2cc0f61957a5)
3. The rest is just some refactoring/cleanups.

xdustinface commented:
> 1. [209f43f](https://github.com/dashpay/dash/commit/209f43ff073b2c99f3db437d7453564ec0ed1b62) should be reverted cause it violates discard/dustrelay fee definitions

Isn't a partial revert like in 1365db1 enough? This way we can still say 100% sure that the same feerate is used for change dust consideration.

> 2. I think you messed up "finished" conditions ([39a6056](https://github.com/dashpay/dash/commit/39a60562aa95ef2a4dbd6a49e00c2cc0f61957a5))

Ops, good catch! 
> 3. The rest is just some refactoring/cleanups.

Picked them all 👍

xdustinface commented:
@UdjinM6 While reviewing your suggestions more precisely i realized that your refactoring of `GetAmountUsed()` in 2b01a09 introduced a rounding bug. Reason it didn't respect fees in the beginning was to always have only one fee calculation with the total number of bytes. With it respecting fees you need to call `GetFee()` twice with subtotals in `TryAddOutput` and `TryAddOutputs` https://github.com/dashpay/dash/blob/2b01a09c6e8263095f316961a62b8bfa514ec487/src/privatesend/privatesend-util.cpp#L169-L172 First call is in  `GetAmountLeft()` which calls `GetAmountUsed()`  https://github.com/dashpay/dash/blob/2b01a09c6e8263095f316961a62b8bfa514ec487/src/privatesend/privatesend-util.cpp#L198-L206 This twice calling `GetFee()` results in a rounding error due to integer cutoffs here https://github.com/dashpay/dash/blob/2b01a09c6e8263095f316961a62b8bfa514ec487/src/policy/feerate.cpp#L28

Lets assume `GetBytesTotal() == 200` and `nBytesOutput == 34` and the feerate `1234 duff/KB`

With 2b01a09 applied: 
`1234*200/1000 = 246,8` (246 in integer world) 
`1234*34/1000 = 41,956` (41 in integer world)
= 287 duffs fee

With the initial commits: 
`1234*234/1000 = 288,756` (288 in integer world)
= 288 duffs fee

With the force push i dropped 2b01a09,  7c30d5d, f90e626 again. The latter two because they bring no drawbacks but IMO without 2b01a09 it just brings more readability to keep them.. still feel free to complain again if you prefer to have them dropped for some reason. There are also the new commits 94529da888c797465db775d6a422d79a82459893, 3cb7d2d and 7502fcb99e30e88ac87187b224ea26f5bebfebab

UdjinM6 commented:
Makes sense. Still don't see a reason for `CoinControl::m_discard_feerate` though abbb195ff5.

xdustinface commented:
> Makes sense. Still don't see a reason for `CoinControl::m_discard_feerate` though [abbb195](https://github.com/dashpay/dash/commit/abbb195ff529338bb6a416577afc0e163bd06115).

Okaaay... just one more time a bit more detailed in case you didn't get the reasoning i have in mind for it: I can't say for sure that there is no chance for the code down in the deeps of `GetDiscardRate()` to give different results when it's called twice with some time elapsed in between, can you? So for me thats the reason to have `CCoinControl::m_discard_rate` as it makes sure we still use the same feerate in both cases (`CTransactionBuilder::IsDust` and `CWallet::CreateTransaction`). What may happen if the results are different is that `CTransactionBuilder::Commit` fails. Sure, then it tries it one more time but what if there is only one utxo left with a dust change, this continues to happen and the mixing starts to stuck because of it? I know, it sounds not very likely 😆 but at the end it's a potential unexpected scenario we could avoid with few lines of code. I rather prefer to prevent such edge cases if i know about them. But if you are sure that can't happen or if you think this is an acceptable case just give me your 👍 and i will drop it.

UdjinM6 commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,225 @@ void CKeyHolderStorage::ReturnAll()
+    dummyReserveKey(pwalletIn),
+    tallyItem(tallyItemIn)
+{
+    // Generate d feerate which will be used to consider if the remainder is dust and will go into fees or not
```
```suggestion
    // Generate a feerate which will be used to consider if the remainder is dust and will go into fees or not
```

UdjinM6 commented on src/wallet/wallet.cpp:
```diff
@@ -3809,8 +3809,7 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
 
-                    if (IsDust(txout, ::dustRelayFee))
-                    {
+                    if (IsDust(txout, ::dustRelayFee)) {
```
pls revert this to original state to avoid adding unrelated changes

xdustinface replied:
Done in ec4cc32

UdjinM6 commented:
Oh, by 👍 I meant that it makes sense and I understand the reason now 🙈 pls ignore my previous suggestion and drop 5fb050b

xdustinface commented:
> Oh, by 👍 I meant that it makes sense and I understand the reason now 🙈 pls ignore my previous suggestion and drop [5fb050b](https://github.com/dashpay/dash/commit/5fb050b2b272862b60986f7c7bc6354a37882d2f)

Ahhh ops, i took it as the "go drop it thumbs up" 😅

UdjinM6 commented:
Pls see https://github.com/UdjinM6/dash/commits/pr3657 (starting with 703922203ecfa8f18a2c0c9ed853d91ab7c3fc36)

xdustinface commented:
Overall nice suggestions but im not a big fan of 703922203ecfa8f18a2c0c9ed853d91ab7c3fc36 yet, convince me :D 

Why does it matter in this case if the change is not handled automatically? The idea of doing it manually was to have two outputs with the same amount in this case as this imo reduces potential false positive detection even more for the "make collateral amounts tx type".

UdjinM6 commented:
> Why does it matter in this case if the change is not handled automatically?

Added f6f0f8ac72. Try mixing with `--paytxfee=0.001` for example and you'll see that it fails without 7039222 - when feerate is much greater than discard feerate, wallet tries to create new change output in such cases and fails with "not enough funds".

> Also there is no "Make sure its not denominated" anymore for the change if its done automatically as the amount is not set through `getAdjustedAmount`.

Right, it's not really used after 7039222 (wasn't ever needed in `case3`, the smallest denom is 10x larger than `GetCollateralAmount`) and it can be dropped together with `UpdateAmount` ced2939491.

xdustinface commented:
Nice catch with the extra byte due to the compact serialize size 👍 I added some commits related to that to wrap it up a bit and to also respect this when trying to add outputs in `TryAddOutput` and `TryAddOutputs`. Let me know what you think. 

8b19c551fd85840aeadeaf64f04ffe655cd6072f is also a good catch! But i modified it a bit in 05ee734f69eab1355839c75f4cc5264bf4db4064, thoughts?
	
And i picked the others except 703922203ecfa8f18a2c0c9ed853d91ab7c3fc36. But i will have a better look at it and your points again later if i get the chance or tomorrow!

UdjinM6 commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1650,10 +1650,17 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
+            while(true) {
+                // Create an potential output
                 vecOutputs.push_back(nAmount);
+                if (!txBuilder.TryAddOutputs(vecOutputs) || txBuilder.CountOutputs() + vecOutputs.size() >= PRIVATESEND_DENOM_OUTPUTS_THRESHOLD) {
```
```suggestion
                if (!txBuilder.TryAddOutputs(vecOutputs) || txBuilder.CountOutputs() + vecOutputs.size() > PRIVATESEND_DENOM_OUTPUTS_THRESHOLD) {
```
because `pop_back` below would result in `PRIVATESEND_DENOM_OUTPUTS_THRESHOLD - 1` max otherwise

UdjinM6 commented on src/wallet/wallet.cpp:
```diff
@@ -3975,8 +3975,7 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                         // Adding another output can change the serialized size of the vout size
-                        unsigned int nBytesCompactSizeDiff = GetSizeOfCompactSize(txNew.vout.size() + 1) - GetSizeOfCompactSize(0);
-                        unsigned int tx_size_with_change = nBytes + change_prototype_size + nBytesCompactSizeDiff;
+                        unsigned int tx_size_with_change = nBytes + change_prototype_size + GetSizeOfCompactSizeDiff(0, txNew.vout.size() + 1);
```
I think this should actually be
```suggestion
                        unsigned int tx_size_with_change = nBytes + change_prototype_size + GetSizeOfCompactSizeDiff(txNew.vout.size(), txNew.vout.size() + 1);
```
my bad 🙈

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1434,83 +1434,75 @@ bool CPrivateSendClientSession::MakeCollateralAmounts(const CompactTallyItem& ta
+        CTransactionBuilderOutput* out2 = txBuilder.AddOutput();
+
+        // Create two equal outputs from the available value. This adds one duff to the fee if txBuilder.GetAmountLeft() is odd.
+        CAmount nAmountOutputs = getAdjustedAmount(txBuilder.GetAmountLeft() / 2);
```
Why would we want to create two equal outputs at GetCollateralAmount() + x instead of one output at GetCollateralAmount() and one at GetCollateralAmount + 2x? the latter seems preferable to me and more in line with current logic

xdustinface replied:
The idea behind the two equal outputs was just trying to reduce chance of false positives for this case in `TransactionRecord::decomposeTransaction` to best avoid something like #3597. 

Given the input amount is random and therefor `x` is random

- With 2x `GetCollateralAmount() + x` you are doing `amount1 == amount2 && IsCollateralAmount(amount1)` 

- With  `GetCollateralAmount()` and `GetCollateralAmount() + 2x` you are doing `amount1 == GetCollateralAmount() && IsCollateralAmount(amount2)`

For me the first bullet point, especially the `amount1 == amount2` with the random `x` included just looked to me like a better solution to prevent false positives in this case. And as i couldn't see any drawbacks i implemented it like this.

Thoughts?

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1434,83 +1434,75 @@ bool CPrivateSendClientSession::MakeCollateralAmounts(const CompactTallyItem& ta
+        CTransactionBuilderOutput* out = txBuilder.AddOutput();
+        out->UpdateAmount(getAdjustedAmount(txBuilder.GetAmountLeft()));
+
+        assert(CPrivateSend::IsCollateralAmount(out->GetAmount()));
```
These asserts kindof scare me tbh...

xdustinface replied:
Why exactly?

PastaPastaPasta replied:
I could imagine a situation where somehow something gets changed in such a way that it's too much or too little to pass this assert, and this isn't caught in testing. I would rather log an error, and return false than assert and crash the application. You could maybe use something like https://github.com/dashpay/dash/blob/fea0036e9fb6a85b59a9a9f7c7454286e42c7113/src/validation.cpp#L1546 if you want to somewhat gracefully shutdown the application

PastaPastaPasta replied:
thoughts @UdjinM6 ?

PastaPastaPasta replied:
@UdjinM6 I would like your thoughts here

xdustinface replied:
Hm imo its nothing to worry about, we have unit tests and  https://github.com/dashpay/dash/blob/9a7b36ac5ac6b2c92dce77ea7c047201e388fb88/src/privatesend/privatesend-client.cpp#L1442 https://github.com/dashpay/dash/blob/9a7b36ac5ac6b2c92dce77ea7c047201e388fb88/src/privatesend/privatesend-client.cpp#L1460

UdjinM6 replied:
imo it's ok

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+    dummyReserveKey.ReturnKey();
+}
+
+bool CTransactionBuilder::TryAddOutput(CAmount nAmountOutput) const
```
I'd like this to be named something else, like `TestAddOutput` to show that this doesn't actually add an output, it only returns true if you COULD add an output of this size. `TryAddOutput` makes it sound like it'll actually add it if this function returns true

xdustinface replied:
Done in eba5836, your COULD made me naming it `CouldAdd` 😄

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+    return GetAmountLeft(GetAmountInitial(), GetAmountUsed() + nAmountOutput, GetFee(nBytes)) >= 0;
+}
+
+bool CTransactionBuilder::TryAddOutputs(const std::vector<CAmount>& vecOutputAmounts) const
```
same

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+    return false;
+}
+
+CTransactionBuilder::CTransactionBuilder(CWallet* pwalletIn, const CompactTallyItem& tallyItemIn) :
```
All of this CTransactionBuilder stuff really needs unit tests

xdustinface replied:
I agree, it makes sense and i actually thought about doing it initially but i didn't because at the moment i probably wouldn't be able to provide them without a decent delay.. i first need to work myself into our unit test stuff and its best practices, never really looked at it yet and if i do it, i want to do it right and understand all of t and not just do something that its done. I have it on my list and i will come back to it once i get the time. Hope that works! But also for the meantime, feel free to provide some unit tests if you can manage to get them together easier.

PastaPastaPasta replied:
I want to hear Udjin's thoughts... This is a relatively complicated class we are adding, and I'm not sure I'd want this merged without at least basic unit tests

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+unsigned int CTransactionBuilder::GetBytesTotal() const
+{
+    // Adding other outputs can change the serialized size of the vout size hence + GetSizeOfCompactSizeDiff()
+   return nBytesBase + vecOutputs.size() * nBytesOutput + ::GetSizeOfCompactSizeDiff(0, vecOutputs.size());
```
```suggestion
    return nBytesBase + vecOutputs.size() * nBytesOutput + ::GetSizeOfCompactSizeDiff(0, vecOutputs.size());
```

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+    return nullptr;
+}
+
+unsigned int CTransactionBuilder::GetBytesTotal() const
```
wouldn't it be the standard to use `uint32_t` here?

xdustinface replied:
hm seems like `unsigned int` is used like this in many places so at this point i can't see like there is a standard for it. For example `CAmount GetRequiredFee(unsigned int nTxBytes)` receives the result of  `GetBytesTotal()` as parameter.

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+   return nBytesBase + vecOutputs.size() * nBytesOutput + ::GetSizeOfCompactSizeDiff(0, vecOutputs.size());
+}
+
+CAmount CTransactionBuilder::GetAmountLeft(const CAmount nAmountInitial, const CAmount nAmountUsed, const CAmount nFee)
+{
+    return nAmountInitial - nAmountUsed - nFee;
+}
```
I'm not sure I understand the point of this function at this point is it's just a glorified subtraction function...

xdustinface replied:
Jup its just a subtraction helper. https://github.com/dashpay/dash/blob/e7dc9da05c9e6cf6be4173ad65aecbc611cd1992/src/privatesend/privatesend-util.h#L119 imo it improves the readability of e.g.

https://github.com/dashpay/dash/blob/e7dc9da05c9e6cf6be4173ad65aecbc611cd1992/src/privatesend/privatesend-util.cpp#L169-L174

As it wraps the subtrahends in `GetAmountLeft()` so you can just see by reading the wrappers name what the point of the return is. I don't see any drawback by doing it like this as the compiler should be smart enough to take care of it, but i neither will fight for keeping it lol.. so if you see any problems with this, just complain again.

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+        nBytesAdditional = nBytesOutput + GetSizeOfCompactSizeDiff(1);
+    }
+
+    // If the calculated fee does not match the fee returned by CreateTransaction aka if this check fails something is messed!
```
```suggestion
    // If the calculated fee does not match the fee returned by CreateTransaction aka if this check fails something is wrong!
```

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +76,244 @@ void CKeyHolderStorage::ReturnAll()
+        strResult = state.GetRejectReason();
+        return false;
+    }
+
+    fKeepKeys = true;
+
+    strResult = wtx.GetHash().ToString();
+
```
I would say remove a few of these empty lines, probably line 300 at least

xdustinface commented:
**Added** f6f0f8a, ee0e3ee98a3e4e4714b7c9ac5d02060d71ad7807, e42c43f65c21c6c484ead5ba63356408f787dcba
**Dropped** 8d4dcd2e484ec8b2949f027b08fae04b677bc0d8 (moved it into a4620e4), 6801983 (not longer needed if #3668  gets merged), part of dfebc25

@UdjinM6 are you good now with #3668 instead of 7039222?

xdustinface commented:
Rebased on develop + e755205

UdjinM6 reviewed (approved):
Seems to be working as expected (together with #3668 )

ACK

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1600,81 +1581,91 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
+
+        auto countPossibleOutputs = [&](CAmount nAmount) -> int {
+            std::vector<CAmount> vecOutputs;
+            while(true) {
```
```suggestion
            while (true) {
```

PastaPastaPasta commented on src/privatesend/privatesend-util.cpp:
```diff
@@ -70,3 +82,250 @@ void CKeyHolderStorage::ReturnAll()
+    if (nNewAmount > 0 && nAmountDiff <= pTxBuilder->GetAmountLeft()) {
+        if (nAmountDiff != 0) {
+            nAmount = nNewAmount;
+        }
```
I don't understand why this `if` statement is here

xdustinface replied:
Oh good catch :D thats a leftover from an earlier iteration of `UpdateAmount` where an amount update triggered some recalculations which would have been be avoided by that if, see 7814ff3bfd393757c4c014b92d9f90e9da3deda0

PastaPastaPasta commented on src/privatesend/privatesend-client.cpp:
```diff
@@ -1600,81 +1581,91 @@ bool CPrivateSendClientSession::CreateDenominated(CAmount nBalanceToDenominate,
+                    if (fAddFinal && nBalanceToDenominate > 0 && nBalanceToDenominate < nDenomValue) {
+                        fAddFinal = false; // add final denom only once, only the smalest possible one
+                        LogPrint(BCLog::PRIVATESEND, "CPrivateSendClientSession::%s -- 1 - FINAL - nDenomValue: %f, nBalanceToDenominate: %f, nOutputs: %d, %s\n",
+                                                     __func__, (float) nDenomValue / COIN, (float) nBalanceToDenominate / COIN, nOutputs, txBuilder.ToString());
```
This introduces clang-tidy warning "Inside a lambda, '__func__' expands to the name of the function call operator; consider capturing the name of the enclosing function explicitly"

xdustinface replied:
see aacb0562138026b75468af14e73d73566a9ae218

PastaPastaPasta reviewed (changes requested):
see https://github.com/PastaPastaPasta/dash/commits/pr_3657

xdustinface commented:
> see https://github.com/PastaPastaPasta/dash/commits/pr_3657

Picked 87b7431, 2543a4c and bcdcb13 (squashed the 3 typo commits) 👍 

About the others: 

71d49f9bd57dc0937358001f5638f1c506b114dc and 6c34988801d640a91d624a3ab8cc7ae70ab0cf58: does it really bring any improvement / make a difference to use explicit casts there? IMO this just blows it up unnecessarily. We will never reach anything near `INT_MAX` in any of those cases, do we? 

eaeb50e7a61dd6125f9cde76977e76770fed29be: why?

PastaPastaPasta commented:
> > see https://github.com/PastaPastaPasta/dash/commits/pr_3657
> 
> Picked [87b7431](https://github.com/dashpay/dash/commit/87b7431f8eb791bf76290d9cc564a90568a96f35), [2543a4c](https://github.com/dashpay/dash/commit/2543a4c142f78775a6376e137db78fd5f51291c0) and [bcdcb13](https://github.com/dashpay/dash/commit/bcdcb1311b077688fdd42e6b22bcf0e969ce1065) (squashed the 3 typo commits) +1
> 
> About the others:
> 
> [71d49f9](https://github.com/dashpay/dash/commit/71d49f9bd57dc0937358001f5638f1c506b114dc) and [6c34988](https://github.com/dashpay/dash/commit/6c34988801d640a91d624a3ab8cc7ae70ab0cf58): does it really bring any improvement / make a difference to use explicit casts there? IMO this just blows it up unnecessarily. We will never reach anything near `INT_MAX` in any of those cases, do we?
> 
> [eaeb50e](https://github.com/dashpay/dash/commit/eaeb50e7a61dd6125f9cde76977e76770fed29be): why?

https://github.com/dashpay/dash/commit/eaeb50e7a61dd6125f9cde76977e76770fed29be resolves a clang-tidy warning

https://github.com/dashpay/dash/commit/71d49f9bd57dc0937358001f5638f1c506b114dc#diff-7d55c5a40c548772705e07a8dff7a49eR188-R189 L189 specifically resolves a clang-tidy, and the rest imo it's good standard to validate when going from unsigned int -> int or long -> int or really any narrowing conversion, that we aren't having something weird happen.

https://github.com/dashpay/dash/commit/6c34988801d640a91d624a3ab8cc7ae70ab0cf58 resolves clang-tidy warning

xdustinface commented:
Ok, seems like im just not a big fan of some of those clang-tidy recommendations lol.. not the first time i noticed that haha.. and i really wonder if we should always follow them.. especially eaeb50e is one i can't understand at all 😆 anyway, they don't do anything bad so i just picked them.

PastaPastaPasta commented:
> Ok, seems like im just not a big fan of some of those clang-tidy recommendations lol.. not the first time i noticed that haha.. and i really wonder if we should always follow them.. especially [eaeb50e](https://github.com/dashpay/dash/commit/eaeb50e7a61dd6125f9cde76977e76770fed29be) is one i can't understand at all laughing anyway, they don't do anything bad so i just picked them.

see https://stackoverflow.com/a/46292715

xdustinface commented:
> see https://stackoverflow.com/a/46292715

It's not that i didn't get that it doesn't make a difference :D I just don't see the point why clang-tidy prefers to introduce inconsistencies between declaration and definition. I mean

```
static CAmount GetAmountLeft(CAmount, CAmount, CAmount);
```
also works, but you also don't leave variable names out of the declaration only because they are ignored by the compiler and they don't have an impact on the definition, or is this also something clang-tidy recommends?

UdjinM6 commented on src/privatesend/privatesend-util.h:
```diff
@@ -34,4 +36,96 @@ class CKeyHolderStorage
+    CTransactionBuilderOutput(CTransactionBuilderOutput&&) = delete;
+    CTransactionBuilderOutput& operator=(CTransactionBuilderOutput&&) = delete;
+    // Get the scriptPubKey of this output
+    CScript GetScript() { return script; }
```
```suggestion
    CScript GetScript() const { return script; }
```

PastaPastaPasta commented:
> > see https://stackoverflow.com/a/46292715
> 
> It's not that i didn't get that it doesn't make a difference :D I just don't see the point why clang-tidy prefers to introduce inconsistencies between declaration and definition. I mean
> 
> ```
> static CAmount GetAmountLeft(CAmount, CAmount, CAmount);
> ```
> 
> also works, but you also don't leave variable names out of the declaration only because they are ignored by the compiler and they don't have an impact on the definition, or is this also something clang-tidy recommends?

Seems like most of the thought process is the fact that something is const doesn't change how someone is going to call that function, whereas knowing the names is pretty valuable

UdjinM6 reviewed (approved):
utACK

PastaPastaPasta commented on src/privatesend/privatesend-util.h:
```diff
@@ -34,4 +36,96 @@ class CKeyHolderStorage
+    // Needed by CTransactionBuilderOutput::UpdateAmount to lock cs_outputs
+    friend class CTransactionBuilderOutput;
+
+public:
+    CTransactionBuilder(CWallet* pwalletIn, const CompactTallyItem& tallyItemIn);
+    ~CTransactionBuilder();
+    // Check it would be possible to add a single output with the amount nAmount. Returns true if its possible and false if not.
+    bool CouldAddOutput(CAmount nAmountOutput) const;
+    // Check if its possible to add multiple outputs as vector of amounts. Returns true if its possible to add all of them and false if not.
+    bool CouldAddOutputs(const std::vector<CAmount>& vecOutputAmounts) const;
+    // Add an output with the amount nAmount. Returns a pointer to the output if it could be added and nullptr if not due to insufficient amount left.
+    CTransactionBuilderOutput* AddOutput(CAmount nAmountOutput = 0);
+    // Get amount we had available when we started
+    CAmount GetAmountInitial() const { return tallyItem.nAmount; }
+    // Get the amount currently left to add more outputs. Does respect fees.
+    CAmount GetAmountLeft() const { return GetAmountInitial() - GetAmountUsed() - GetFee(GetBytesTotal()); }
+    // Check if an amounts should be considered as dust
+    bool IsDust(CAmount nAmount) const;
+    // Get the total number of added outputs
+    int CountOutputs() const { return vecOutputs.size(); }
+    // Create and Commit the transaction to the wallet
+    bool Commit(std::string& strResult);
+    // Convert to a string
+    std::string ToString() const;
+
+private:
+    // Clear the output vector and keep/return the included keys depending on the value of fKeepKeys
+    void Clear();
+    // Get the total number of bytes used already by this transaction
+    unsigned int GetBytesTotal() const;
+    // Helper to calculate static amount left by simply subtracting an used amount and a fee from a provided initial amount.
+    static CAmount GetAmountLeft(CAmount nAmountInitial, CAmount nAmountUsed, CAmount nFee);
+    // Get the amount currently used by added outputs. Does not include fees.
+    CAmount GetAmountUsed() const;
+    // Get fees based on the number of bytes and the feerate set in CoinControl.
+    // NOTE: To get the total transaction fee this should only be called once with the total number of bytes for the transaction to avoid
+    // calling CFeeRate::GetFee multiple times with subtotals as this may add rounding errors with each further call.
+    CAmount GetFee(unsigned int nBytes) const;
+    // Helper to get GetSizeOfCompactSizeDiff(vecOutputs.size(), vecOutputs.size() + nAdd)
+    int GetSizeOfCompactSizeDiff(size_t nAdd) const;
```
kindof nit: can you make these comments doxygen? `///`?

xdustinface replied:
done in c30df0d

PastaPastaPasta commented:
I think I'm ready to ACK once those comments become doxygen

PastaPastaPasta reviewed (approved):
ACK, code review, as well as mixed a good bit on testnet
